### PR TITLE
Add reverse color option to text diff editor

### PR DESCRIPTION
**Pull request by kisantia**
This changeset adds the color reversing option as an optional parameter for the diff editor. This is needed for the schema compare extension since it uses the text diff editor and needs the colors to be reversed.

image

image

image

### CODE REVIEW

**Comment by anthonydresser** [comment]
I thought the plan was to keep the current diff coloring scheme and accept that the color scheme won't be the same as in ssdt? @yualan @vickyharp @kenvanhyning

File: src/vs/editor/browser/widget/diffEditorWidget.ts [hunk 1]
```diff
@@ -1203,19 +1205,53 @@ abstract class DiffEditorWidgetStyle extends Disposable implements IDiffEditorWi
...
+
+		let originalDecorations: IEditorDiffDecorations;
+		let modifiedDecorations: IEditorDiffDecorations;
 		let zones = this._getViewZones(lineChanges, originalWhitespaces, modifiedWhitespaces, originalEditor, modifiedEditor, renderIndicators);
 
-		// Get decorations & overview ruler zones
-		let originalDecorations = this._getOriginalEditorDecorations(lineChanges, ignoreTrimWhitespace, renderIndicators, originalEditor, modifiedEditor);
-		let modifiedDecorations = this._getModifiedEditorDecorations(lineChanges, ignoreTrimWhitespace, renderIndicators, originalEditor, modifiedEditor);
+		// if we need to reverse coloring
+		if (reverse) {
```
Comment (udeeshagautam) [inline src/vs/editor/browser/widget/diffEditorWidget.ts]: Should we take logic inside reverse out in separate private function for readability?

File: src/vs/editor/browser/widget/diffEditorWidget.ts [hunk 2]
```diff
@@ -61,7 +61,7 @@ interface IEditorsZones {
 }
 
 interface IDiffEditorWidgetStyle {
-	getEditorsDiffDecorations(lineChanges: editorCommon.ILineChange[], ignoreTrimWhitespace: boolean, renderIndicators: boolean, originalWhitespaces: IEditorWhitespace[], modifiedWhitespaces: IEditorWhitespace[], originalEditor: editorBrowser.ICodeEditor, modifiedEditor: editorBrowser.ICodeEditor): IEditorsDiffDecorationsWithZones;
+	getEditorsDiffDecorations(lineChanges: editorCommon.ILineChange[], ignoreTrimWhitespace: boolean, renderIndicators: boolean, originalWhitespaces: IEditorWhitespace[], modifiedWhitespaces: IEditorWhitespace[], originalEditor: editorBrowser.ICodeEditor, modifiedEditor: editorBrowser.ICodeEditor, reverse?: boolean): IEditorsDiffDecorationsWithZones;
```
Comment (kburtram) [inline src/vs/editor/browser/widget/diffEditorWidget.ts]: Please add a `// {{SQL CARBON EDIT}} comment to each distinct block of code changed under the `vs` directory.  This is needed to help when merging in VS Code source code refreshes so we can easily identify our patches.

Also, can you please try to refactor this change to minimize the number of lines of code changed and distinct code blocks added under `vs` directory.  Perhaps by creating a help class under `sql` and then just adding method calls from under `vs`.  

Basically we want to keep changes to vs source code minimal and narrowly scoped.

Comment (kisantia) [reply src/vs/editor/browser/widget/diffEditorWidget.ts]: I moved some of the code to a new class diffEditorHelper.ts in sql/editor/browser. Is this the right place to put it?

File: src/vs/editor/browser/widget/diffEditorWidget.ts (hunk 1 above)
Comment (kisantia) [reply src/vs/editor/browser/widget/diffEditorWidget.ts]: Moved the line change reversing to a helper class in diffEditorHelper.ts

File: src/sql/editor/browser/diffEditorHelper.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import * as editorCommon from 'vs/editor/common/editorCommon';
+
+export class DiffEditorHelper {
```
Comment (Charles-Gagnon) [inline src/sql/editor/browser/diffEditorHelper.ts]: You can just export the functions directly - much simpler and cleaner

File: src/sql/editor/browser/diffEditorHelper.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import * as editorCommon from 'vs/editor/common/editorCommon';
```
Comment (Charles-Gagnon) [inline src/sql/editor/browser/diffEditorHelper.ts]: Don't forget the copyright block

File: src/sql/editor/browser/diffEditorHelper.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+export class DiffEditorHelper {
+	public static reverseLineChanges(lineChanges: editorCommon.ILineChange[]): editorCommon.ILineChange[] {
```
Comment (Charles-Gagnon) [inline src/sql/editor/browser/diffEditorHelper.ts]: Doc comments please
